src/language/visitor.ts

```diff
@@ -1,28 +1,47 @@
 import type * as babel from '@babel/core';
 import * as parser from '@babel/parser';
 import traverse from '@babel/traverse';
+import { Connection } from 'vscode-languageserver/node';
+import * as util from 'util';
 
 const PLACEHOLDER = 'TRIGGER_CHARACTER';
 
-export type CompletionState = {
+export interface VisitorSelection {
+  start: { line: number, character: number };
+  end: { line: number, character: number };
+}
+
+export interface VisitorTextAndSelection {
+  textFromEditor: string;
+  selection: VisitorSelection;
+}
+
+export interface CompletionState {
   databaseName: string | null;
   collectionName: string | null;
+  isObject: boolean;
+  isArray: boolean;
   isObjectKey: boolean;
   isShellMethod: boolean;
   isUseCallExpression: boolean;
   isDbCallExpression: boolean;
   isCollectionName: boolean;
   isAggregationCursor: boolean;
   isFindCursor: boolean;
-};
+}
 
 export class Visitor {
   _state: CompletionState;
-  _position: { line: number; character: number };
+  _selection: VisitorSelection;
+  _connection: Connection;
```

Comment:
This is only used for printing error messages, right? Would it makes sense to decouple this a bit and keep this `Visitor` class independent of VSCode specifics, and instead pass in e.g. a logging callback or only the `console` object?